File: solutions/c/rotational-cipher/1/rotational_cipher.c
#include "rotational_cipher.h"

#include <stdlib.h>
#include <string.h>

char *rotate(const char *text, int shift_key) {
    int len = strlen(text);
    char *res = calloc(len + 1, sizeof(char));

    for (int i = 0; i < len; i++) {
        if (('a' <= text[i] && text[i] <= 'z') ||
            ('A' <= text[i] && text[i] <= 'Z')) {
            char base = 'a' <= text[i] ? 'a' : 'A';
            res[i] = (((text[i] - base) + shift_key) % 26) + base;
        } else
            res[i] = text[i];
    }

    return res;
}

File: solutions/java/rotational-cipher/1/src/main/java/RotationalCipher.java
import java.util.stream.Collectors;

class RotationalCipher {
    private final int key;

    RotationalCipher(int shiftKey) {
        this.key = shiftKey;
    }

    String rotate(String data) {
        return data.chars().mapToObj((c) -> Character.toString(charRot(c, key))).collect(Collectors.joining());
    }

    private static int charRot(int c, int key) {
        if (!Character.isLetter(c))
            return (int) c;
        char base = Character.isUpperCase(c) ? 'A' : 'a';
        return (c - base + key) % 26 + base;
    }
}
